### PR TITLE
Added multipart/form-data support

### PR DESCRIPTION
**Pull request by MrToino**
Added a complete initial multipart support, from frontend to backend services.

### CODE REVIEW

File: intmodules/parser.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    },
+    "requied": []
+}
+
+
+def parse_request(request: tuple) -> dict:
+    """Validates request json in format and schema"""
+
+    if len(request) == 2:
+        file_part, options_part = (part.content.decode(part.encoding) for part in request)
```
Comment (luisferreira32) [inline intmodules/parser.py:23]: Do we even have guaranteed order? Is multipart form data not a key-value map?

File: intmodules/parser.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    elif len(request) == 3:
+        file_part, options_part, _ = (part.content.decode(part.encoding) for part in request)
+    else:
+        raise InvalidRequestError
+
+    data = read_csv(StringIO(file_part), sep=",", index_col=0)
```
Comment (luisferreira32) [inline intmodules/parser.py:29]: If the file in the end is not a csv exactly as you're expecting, will this data be anything meaningful? Will the function throw an exception? i.e., do we want to do any extra validation over here?

File: intmodules/preprocessor.py [hunk 3]
```diff
@@ -1,9 +1,29 @@
...
+def check_index_duplicates(data):
+    """Check if data index has duplicates"""
+    index = data.index.values
+    if not len(index) == len(set(index)):
+        raise DuplicatedEntryError
+
+
+def preprocess_data(data):
+    """Validates input data"""
+    check_null_values(data)
```
Comment (luisferreira32) [inline intmodules/preprocessor.py:26]: This three functions (`check_null_values`, `check_index_duplicates`, and `check_numeric_values`) are two liners, you win nothing doing an abstraction, and I would deeply advise against using them in any service (would have to make you jump between files to actually know what's happening) - making this three helper functions absolutely useless besides their use in the `preprocess_data` function. So , in short, this `preprocess_data` should have the content of the functions themselves, and the three two liners should not exist as separate functions.

File: services/fitter/.env [hunk 4]
```diff
@@ -1,2 +1,2 @@
 SERVER_HOSTNAME="localhost"
-SERVER_PORT="8080"
\ No newline at end of file
+SERVER_PORT=8081
```
Comment (luisferreira32) [inline services/fitter/.env:2]: Thanks for the correction :+1:

File: services/fitter/server.py [hunk 5]
```diff
@@ -15,7 +14,8 @@ def do_POST(self):
         request = self.rfile.read(length)
 
         try:
-            response = service(request)
+            parsed_request = self._multipart_parser(request)
+            response = service(parsed_request)
         except ReadingBase64Error:
```
Comment (luisferreira32) [inline services/fitter/server.py:19]: I guess this exception should jump out?

File: services/fitter/server.py [hunk 6]
```diff
@@ -44,15 +44,15 @@ def do_POST(self):
...
 if __name__ == "__main__":
     filename, level = "fitter.log", 10
-    config_logger(filename=filename)
-    logger.debug("Initialized logger for plotter service at {} with level {}.".format(filename, level))
+    # config_logger(filename=filename)
+    # logger.debug("Initialized logger for plotter service at {} with level {}.".format(filename, level))
 
     hostName = os.getenv('SERVER_HOSTNAME', 'localhost')
-    serverPort = os.getenv('SERVER_PORT', '8081')
+    serverPort = int(os.getenv('SERVER_PORT', 8081))
```
Comment (luisferreira32) [inline services/fitter/server.py:51]: This is already like this in main... are you sure you've rebased your branch properly?

File: services/plotter/server.py [hunk 7]
```diff
@@ -31,8 +27,8 @@ def do_POST(self):
         except ValueError:
             response = "Bad Request: ValueError"
             self._return_400(response)
-        except UnicodeDecodeError:
-            response = "Bad Request: UnicodeDecodeError"
+        except Exception:
+            response = "Oops: something went wrong"
             self._return_400(response)
```
Comment (luisferreira32) [inline services/plotter/server.py:32]: Should return 500 here

File: tests/test.json [hunk 8]
```diff
@@ -1 +1 @@
-{"Filename" : "test", "ContentB64" : "eCx5CjEsMQoyLDQKMyw5CjQsMTYKNSwyNQo2LDM2CjcsNDkKOCw2NA=="}
\ No newline at end of file
+{"Filename" : "test", "ContentB64" : "eCx5CjEsMQoyLDQKMyw5CjQsMTYKNSwyNQo2LDM2CjcsNDkKOCw2NA==", "Configurations" : ""}
```
Comment (luisferreira32) [inline tests/test.json:1]: Do we need this file even? I'd say it's time to yeet

File: application/src/components/FitterTab.tsx [hunk 9]
```diff
@@ -23,11 +23,16 @@ const FitterTab: React.FC<FitterTabProps> = ({ submitter, setPlot }) => {
     setFittingFunc(fittingFunc);
   };
 
+  const default_json = JSON.stringify({
+    title: 'Title tities',
```
[the remark this answers is not in the document]
Comment (MrToino) [reply application/src/components/FitterTab.tsx:27]: One: oops, i will change indeed

Two: This default_json will later be removed, as the json options object will be built from tab forms. This was the simplest way I found to just set up a local json for testing.

File: intmodules/parser.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        "title": {
+            "type": "string"
+        }
+    },
+    "requied": []
+}
+
+
+def parse_request(request: tuple) -> dict:
+    """Validates request json in format and schema"""
```
[the remark this answers is not in the document]
Comment (MrToino) [reply intmodules/parser.py:20]: Docstring was not updated on this PR.

File: intmodules/parser.py (hunk 1 above)
Comment (MrToino) [reply intmodules/parser.py:23]: Yes, i verified the request is ordered accordingly to the front-end building request. Nevertheless this is a first approach and later on I will be implementing the same not as a function of the order but regarding to the headers.

File: intmodules/parser.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+    if len(request) == 2:
+        file_part, options_part = (part.content.decode(part.encoding) for part in request)
+    elif len(request) == 3:
```
[the remark this answers is not in the document]
Comment (MrToino) [reply intmodules/parser.py:24]: There is no need to have this functions duplicated. One can simply pass the engine (plotter, fitter, ...) as an argument on the request. I'd say it is more simple but right, maybe later .we should indeed have an utils package at the microservice level.

File: services/fitter/server.py (hunk 6 above)
Comment (MrToino) [reply services/fitter/server.py:51]: Yeah, somehow there were some issues that were supposed to be solved but were not. Let it pass

File: services/plotter/server.py (hunk 7 above)
Comment (MrToino) [reply services/plotter/server.py:32]: Yes, 500 not yet implemented. Errors and exceptions will be revised on future PR.

File: tests/test.json (hunk 8 above)
Comment (MrToino) [reply tests/test.json:1]: No, we no longer need it. The invoker itself no longer make sense and i will be cleaning the tests and the invoker soon.

File: application/src/components/FitterTab.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (luisferreira32) [reply application/src/components/FitterTab.tsx:27]: But I did tell you exactly how you should keep the default json for testing: `useState` (something similar to the file, if you look at the top of the file). Not saying you have to manipulate it yet, but keep it already as a state of the component

File: application/src/api/fitter.ts [hunk 12]
```diff
@@ -6,8 +6,8 @@ export const submitFitRequest: formSubmitter = async (formData: FormData) => {
   const response = await fetch(fitterApi, {
     method: "POST",
     headers: {
-      "Content-type": "multipart/form-data",
-      "Access-Control-Allow-Origin": "*", // FIXME: this is to go around localhost issues
+      // "Content-type": "multipart/form-data",
```
[the remark this answers is not in the document]
Comment (MrToino) [reply application/src/api/fitter.ts:9]: Ok, will do.

File: intmodules/parser.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (luisferreira32) [reply intmodules/parser.py:20]: But the function logic was, so, either delete the docstring or update it 😛

File: intmodules/parser.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (luisferreira32) [reply intmodules/parser.py:24]: Simply passing the engine (i.e., service) to do some logic inside the function is a major red flag for a common package. A common package should treat common logic (e.g., auth, logs, data util functions, etc.). This is to validate the request itself, so it's a service based logic

File: application/src/components/FitterTab.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (MrToino) [reply application/src/components/FitterTab.tsx:27]: Sorry, i don't know much about frontend. Will do as you tell.

File: intmodules/exceptions.py [hunk 13]
```diff
@@ -15,3 +9,8 @@ class DuplicatedEntryError(Exception):
     def __init__(self):
         super().__init__(self)
 
+
+class InvalidRequestError(Exception):
```
[the remark this answers is not in the document]
Comment (MrToino) [reply intmodules/exceptions.py:13]: Yes, that is a good idea in fact. In a next PR we will introduce a new implementation of the errors/exceptions.

File: intmodules/preprocessor.py (hunk 3 above)
Comment (MrToino) [reply intmodules/preprocessor.py:26]: Solved.

File: intmodules/parser.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (MrToino) [reply intmodules/parser.py:20]: Anyway, solved

**Review by luisferreira32 — APPROVED** [review]
Still a lot to work on, but a nice baby step :+1:

**Pull request merged**